src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts

```diff
@@ -1195,97 +1188,125 @@ export class DirtyDiffModel extends Disposable {
+				return Promise.resolve([]); // disposed
+			}
 
-				const ignoreTrimWhitespaceSetting = this.configurationService.getValue<'true' | 'false' | 'inherit'>('scm.diffDecorationsIgnoreTrimWhitespace');
-				const ignoreTrimWhitespace = ignoreTrimWhitespaceSetting === 'inherit'
-					? this.configurationService.getValue<boolean>('diffEditor.ignoreTrimWhitespace')
-					: ignoreTrimWhitespaceSetting !== 'false';
+			if (originalURIs.some(quickDiff => !this.editorWorkerService.canComputeDirtyDiff(quickDiff.originalResource, this._model.resource))) {
+				return Promise.resolve([]); // Files too large
+			}
 
-				return this.editorWorkerService.computeDirtyDiff(originalURI, this._model.resource, ignoreTrimWhitespace);
-			});
+			const ignoreTrimWhitespaceSetting = this.configurationService.getValue<'true' | 'false' | 'inherit'>('scm.diffDecorationsIgnoreTrimWhitespace');
+			const ignoreTrimWhitespace = ignoreTrimWhitespaceSetting === 'inherit'
+				? this.configurationService.getValue<boolean>('diffEditor.ignoreTrimWhitespace')
+				: ignoreTrimWhitespaceSetting !== 'false';
+
+			const allDiffs: { change: IChange; labels: string[]; uri: URI }[] = [];
+			for (const quickDiff of originalURIs) {
+				const dirtyDiff = await this.editorWorkerService.computeDirtyDiff(quickDiff.originalResource, this._model.resource, ignoreTrimWhitespace);
+				if (dirtyDiff) {
+					for (const diff of dirtyDiff) {
+						if (diff) {
+							allDiffs.push({ change: diff, labels: [quickDiff.label], uri: quickDiff.originalResource });
+						}
+					}
+				}
+			}
+			const sorted = allDiffs.sort((a, b) => compareChanges(a.change, b.change));
+			const reduced = [];
+			for (const diff of sorted) {
+				if (reduced.length === 0) {
+					reduced.push(diff);
+				} else {
+					const comparedChanges = compareChanges(reduced[reduced.length - 1].change, diff.change);
+					if (comparedChanges === 0) {
+						reduced[reduced.length - 1].labels.push(...diff.labels);
+					} else {
+						reduced.push(diff);
+					}
+				}
+			}
+			return reduced;
 		});
 	}
 
-	private getOriginalURIPromise(): Promise<URI | null> {
-		if (this._originalURIPromise) {
-			return this._originalURIPromise;
+	private getQuickDiffsPromise(): Promise<QuickDiff[]> {
+		if (this._quickDiffsPromise) {
+			return this._quickDiffsPromise;
 		}
 
-		this._originalURIPromise = this.getOriginalResource().then(originalUri => {
+		this._quickDiffsPromise = this.getOriginalResource().then(async (quickDiffs) => {
 			if (this._disposed) { // disposed
-				return null;
+				return [];
 			}
 
-			if (!originalUri) {
-				this._originalResource = null;
-				this._originalModel = null;
-				return null;
+			if (quickDiffs.length === 0) {
+				this._quickDiffs = null;
+				this._originalModels.clear();
+				return [];
 			}
 
-			if (this._originalResource?.toString() === originalUri.toString()) {
-				return originalUri;
+			if (this._quickDiffs?.toString() === quickDiffs.toString()) {
```

Comment:
Isn't `_quickDiffs` an array of objects? This is probably not the best way to compare them. See [`arrays.ts#equals`](https://github.com/microsoft/vscode/blob/4d882322ca0dc62625ebbd5076ebb82a1d250a42/src/vs/base/common/arrays.ts#L27).